go/vt/wrangler/reparent.go

```diff
@@ -406,52 +408,202 @@ func (wr *Wrangler) plannedReparentShardLocked(ctx context.Context, ev *events.R
+		defer promoteCancel()
+		rp, err := wr.tmc.PromoteSlave(promoteCtx, masterElectTabletInfo.Tablet)
 		if err != nil {
-			return fmt.Errorf("can't get current replication position of master: %v", err)
+			return vterrors.Wrapf(err, "failed to promote %v to master", masterElectTabletAliasStr)
 		}
 		reparentJournalPos = rp
-	} else {
-		// If the current master is not the one we want, demote the old master.
-		// Note that since the shard record is eventually consistent, we might be
-		// demoting a former master that was already demoted.
+	} else if topoproto.TabletAliasEqual(currentMaster.Alias, masterElectTabletAlias) {
+		refreshCtx, refreshCancel := context.WithTimeout(ctx, *topo.RemoteOperationTimeout)
+		defer refreshCancel()
+
+		// The master is already the one we want according to its tablet record.
+		// Refresh it to make sure the tablet has read its record recently.
+		if err := wr.tmc.RefreshState(refreshCtx, masterElectTabletInfo.Tablet); err != nil {
+			return vterrors.Wrapf(err, "failed to RefreshState on current master %v", masterElectTabletAliasStr)
+		}
 
-		oldMasterTabletInfo, ok := tabletMap[topoproto.TabletAliasString(shardInfo.MasterAlias)]
-		if !ok {
-			return fmt.Errorf("old master tablet %v is not in the shard", topoproto.TabletAliasString(shardInfo.MasterAlias))
+		// Then get the position so we can try to fix replicas (below).
+		rp, err := wr.tmc.MasterPosition(refreshCtx, masterElectTabletInfo.Tablet)
+		if err != nil {
+			return vterrors.Wrapf(err, "failed to get replication position of current master %v", masterElectTabletAliasStr)
 		}
+		reparentJournalPos = rp
+	} else {
+		// There is already a master and it's not the one we want.
+		oldMasterTabletInfo := currentMaster
 		ev.OldMaster = *oldMasterTabletInfo.Tablet
 
-		// Demote the old master and get its replication position.
-		wr.logger.Infof("demote current master %v", shardInfo.MasterAlias)
+		// Before demoting the old master, check replication on the candidate
+		// master to make sure it has a chance of catching up.
+		statusCtx, statusCancel := context.WithTimeout(ctx, *topo.RemoteOperationTimeout)
+		defer statusCancel()
+		status, err := wr.tmc.SlaveStatus(statusCtx, masterElectTabletInfo.Tablet)
+		if err != nil {
+			return vterrors.Wrapf(err, "can't get replication status on master-elect %v", masterElectTabletAliasStr)
+		}
+		if !status.SlaveIoRunning || !status.SlaveSqlRunning {
+			return vterrors.Errorf(vtrpcpb.Code_FAILED_PRECONDITION, "replication not running on master-elect %v; replication must be healthy to perform planned reparent", masterElectTabletAliasStr)
+		}
+		// Check if it's behind by a small enough amount that it's likely to
+		// catch up before we time out. This assumes a replica can work through
+		// its backlog at approximately the same rate that the transactions
+		// happened on the master.
+		if float64(status.SecondsBehindMaster) >= waitReplicasTimeout.Seconds() {
```

Comment:
One concern with piggy-backing the replication lag timeout on waitReplicasTimeout is that it doesn't allow people to opt out of this check. We are arguing that this is OK because a replica is unlikely to catch up during that time.

SecondsBehindMaster is in fact not reliable and there are cases when it doesn't get updated often enough by mysql and is actually reported as much higher than the real lag. This seems to happen in precisely the situation we are creating here - where there are no more writes to master.
See last comment in https://github.com/vitessio/vitess/pull/5000#issuecomment-530656436

We could leave this as implemented or consider adding another flag (`-skip_replication_lag_check`) to the PRS command.

Reply:
Good catch. What about making a flag to customize the SecondsBehindMaster threshold, instead of an on/off thing?

Reply:
That is what I had originally planned to implement for #4700, so I'll vote yes on that.